lib/bandit/http1/handler.ex

```diff
@@ -29,21 +29,14 @@ defmodule Bandit.HTTP1.Handler do
   end
 
   defp build_conn(req) do
-    case Adapter.read_headers(req) do
-      {:ok, headers, method, request_target, req} ->
-        %{address: remote_ip} = Adapter.get_peer_data(req)
-
-        # Parse a string to build a URI struct. This is quite a hack. In general, canonicalizing
-        # URIs is a delicate process & rather than building a half-baked implementation here it's
-        # better to leave a simple and ugly hack in place so that future improvements are obvious.
-        # Future paths here are discussed at https://github.com/elixir-plug/plug/issues/948)
-        {"host", host} = List.keyfind(headers, "host", 0, {"host", nil})
-        scheme = if Adapter.secure?(req), do: :https, else: :http
-
-        uri = build_uri(scheme, host, request_target)
-
-        {:ok, Plug.Conn.Adapter.conn({Adapter, req}, method, uri, remote_ip, headers)}
-
+    with {:ok, headers, method, request_target, req} <- Adapter.read_headers(req),
+         {:ok, scheme} <- determine_scheme(request_target, req),
+         {:ok, host, port} <- determine_host_and_port(request_target, headers, req),
+         {:ok, path, query} <- determine_path_and_query(request_target),
+         uri <- %URI{scheme: scheme, host: host, port: port, path: path, query: query},
```

Comment:
Since I'm bringing up this in the other PR, could be `=` since it will always match and never fall through...

```suggestion
         uri = %URI{scheme: scheme, host: host, port: port, path: path, query: query},
```

Reply:
Will do manually since I can make the same optimization on the following line as well!